### PR TITLE
[IMP] web: use dialog service on action service

### PR DESCRIPTION
**Pull request by jpp-odoo**

### CODE REVIEW

File: addons/web/static/src/core/dialog/dialog_container.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+export class DialogContainer extends Component {
+    setup() {
+        this.dialogs = {};
+        const { bus } = useService("dialog");
+        bus.on("UPDATE", this, (dialogs) => {
+            this.dialogs = dialogs;
+            this.render();
+        });
+        bus.on("CLOSE", this, (dialogs, id) => {
+            this.dialogs = dialogs; // Overkill ???
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog_container.js]: I think so (?)

File: addons/web/static/src/core/dialog/dialog_service.js [hunk 2]
```diff
@@ -1,76 +1,34 @@
...
+                id,
+                class: dialogClass,
+                props,
+                options,
+            };
+            bus.trigger("UPDATE", dialogs);
+            return id;
+        }
+        function close(id) {
+            bus.trigger("CLOSE", dialogs, id);
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog_service.js]: dialogs isn't updated?

File: addons/web/static/src/webclient/actions/action_container.js [hunk 3]
```diff
@@ -47,17 +24,11 @@ export class ActionContainer extends Component {
...
-        this.dialog = {};
-        this.render();
-    }
 }
 ActionContainer.components = { ActionDialog };
 ActionContainer.template = tags.xml`
     <t t-name="web.ActionContainer">
       <div class="o_action_manager">
         <t t-if="main.Component" t-component="main.Component" t-props="main.componentProps" t-key="main.id"/>
-        <ActionDialog t-if="dialog.id" t-props="dialog.props" t-key="dialog.id" t-on-dialog-closed="onDialogClosed"/>
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_container.js:61]: I love the changes in this file :+1: 

this.main could probably be renamed now

File: addons/web/static/src/webclient/actions/action_service.js [hunk 4]
```diff
@@ -535,12 +539,16 @@ function makeActionManager(env) {
...
-                onClose: options.onClose,
-            });
+
+            const { onClose } = dialog;
+            if (dialog.dialogId) {
+                env.services.dialog.close(dialog.dialogId);
+            }
+            const dialogId = env.services.dialog.open(ActionDialog, actionDialogProps);
+            dialog = {
+                dialogId: dialogId,
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js]:
```suggestion
                id: dialogId,
```

File: addons/web/static/src/webclient/actions/action_service.js [hunk 5]
```diff
@@ -527,6 +530,7 @@ function makeActionManager(env) {
         ControllerComponent.Component = controller.Component;
 
         if (action.target === "new") {
+            cleanDomFromBootstrap();
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:533]: I need explanations

File: addons/web/static/src/webclient/actions/action_service.js [hunk 6]
```diff
@@ -894,13 +902,20 @@ function makeActionManager(env) {
...
 
     async function _executeCloseAction(params = {}) {
-        const closingProms = [];
-        env.bus.trigger("ACTION_MANAGER:UPDATE", {
-            type: "CLOSE_DIALOG",
-            closingProms,
-            ...params,
-        });
-        await Promise.all([dialogCloseProm].concat(closingProms.map((fn) => fn())));
+        cleanDomFromBootstrap();
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:904]: same here

File: addons/web/static/src/webclient/actions/action_service.js [hunk 7]
```diff
@@ -894,13 +902,20 @@ function makeActionManager(env) {
...
+            env.services.dialog.close(dialog.dialogId);
+        } else {
+            onClose = params.onClose;
+        }
+        if (onClose) {
+            await onClose(params.onCloseInfo);
+        }
+        dialog = {};
+
+        return dialogCloseProm;
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:917]: is this still necessary, or could we simply use the promise returned by the call to dialog.close?

File: addons/web/static/src/webclient/actions/action_service.js [hunk 8]
```diff
@@ -894,13 +902,20 @@ function makeActionManager(env) {
...
+        cleanDomFromBootstrap();
+        let onClose;
+        if (dialog.dialogId) {
+            onClose = dialog.onClose;
+            env.services.dialog.close(dialog.dialogId);
+        } else {
+            onClose = params.onClose;
+        }
+        if (onClose) {
+            await onClose(params.onCloseInfo);
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:914]: not sure if we synchronize stuff correctly here:
 - ask for dialog close
 -  directly call onClose (and wait for it)
 - return a promise resolved when the dialog is closed (could be closed earlier actually, but we wait for onClose)

File: addons/web/static/tests/legacy/services/service_provider_adapter_notification_tests.js [hunk 9]
```diff
@@ -22,6 +26,8 @@ QUnit.module("Service Provider Adapter Notification", (hooks) => {
   QUnit.test(
     "can display and close a sticky danger notification with a title (legacy)",
     async function (assert) {
+      clearRegistryWithCleanup(mainComponentRegistry);
+      mainComponentRegistry.add("NotificationContainer", NotificationContainer);
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/services/service_provider_adapter_notification_tests.js]: Not sure to understand the reason of this change

**Review by aab-odoo — COMMENTED** [review]
Cool PR :)

File: addons/web/static/src/core/dialog/dialog_service.js [hunk 10]
```diff
@@ -1,76 +1,34 @@
...
-    `;
-
-registry.category("main_components").add("DialogContainer", DialogContainer);
-
 export const dialogService = {
-    start(env) {
+    start() {
         const bus = new EventBus();
+        let dialogId = 0;
+        let dialogs = [];
```
Comment (kebeclibre) [inline addons/web/static/src/core/dialog/dialog_service.js]: i think there is a choice to make.
either the service keeps the list of dialogs, or the container does. There is a mix of both here.

My prefered solution is:
- the service is in charge of the ID (always incrementing, so always unique)
- the container is in charge of maintaining the iterable containing the current opened dialogs

File: addons/web/static/src/webclient/actions/action_container.js [hunk 11]
```diff
@@ -10,34 +10,11 @@ const { Component, tags } = owl;
 export class ActionContainer extends Component {
     setup() {
         this.main = {};
-        this.dialog = {};
         this.env.bus.on("ACTION_MANAGER:UPDATE", this, (info) => {
             switch (info.type) {
                 case "MAIN":
```
Comment (kebeclibre) [inline addons/web/static/src/webclient/actions/action_container.js]: there are is no other case now....

File: addons/web/static/src/webclient/actions/action_service.js (hunk 7 above)
Comment (kebeclibre) [reply addons/web/static/src/webclient/actions/action_service.js:917]: dialogcloseProm is resolved on the willUnmount of the dialog.....

File: addons/web/static/src/core/dialog/dialog_container.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        throw error;
+    }
+}
+ErrorHandler.template = tags.xml`<t t-component="props.dialog.class" t-props="props.dialog.props" />`;
+
+export class DialogContainer extends Component {
+    setup() {
+        this.dialogs = {};
+        const { bus } = useService("dialog");
+        bus.on("UPDATE", this, (dialogs) => {
```
Comment (mcm-odoo) [inline addons/web/static/src/core/dialog/dialog_container.js:20]: I come late and I didn't follow the dev but why not use the `useBus` hook?

File: addons/web/static/src/core/dialog/dialog_container.js [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        bus.on("CLOSE", this, (id) => {
+            this.doCloseDialog(id);
+        });
+    }
+
+    onDialogClosed(id) {
+        this.doCloseDialog(id);
+    }
+
+    doCloseDialog(id) {
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog_container.js]:
```suggestion
    closeDialog(id) {
```

File: addons/web/static/src/core/dialog/dialog_container.js [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        delete this.dialogs[id];
+        this.render();
+    }
+
+    errorCallBack(id) {
+        return () => this.doCloseDialog(id);
+    }
+
+    destroy() {
```
Comment (aab-odoo) [inline addons/web/static/src/core/dialog/dialog_container.js]: __destroy()

**Comment by aab-odoo** [comment]
robodoo r+

**Pull request closed without merging**